commit 70ab25517c62b32463e3bb502d505b54df45fe62 (refs/changes/69/4969/2)
Author: Tim Foster <tim.foster@joyent.com>
Date:   2018-10-17T17:13:30+01:00 (12 months ago)
    
    TRITON-853 cloudapi should allow creation of instances with delegated datasets

diff --git a/lib/do_instance/do_create.js b/lib/do_instance/do_create.js
index cbedb35..2bebbb2 100644
--- a/lib/do_instance/do_create.js
+++ b/lib/do_instance/do_create.js
@@ -299,6 +299,19 @@ function do_create(subcmd, opts, args, cb) {
                     createOpts.firewall_enabled = opt.value;
                 } else if (opt.key === 'deletion_protection') {
                     createOpts.deletion_protection = opt.value;
+                } else if (opt.key === 'delegate_dataset') {
+                    var allowed_delegate_vals = ['true', 'false', 'safe'];
+                    if (allowed_delegate_vals.indexOf(
+                        opt.value.toLowerCase()) === -1) {
+                        throw new errors.UsageError(
+                            '--delegate-dataset value must be one of: ' +
+                            '[true, false, safe]');
+                    }
+                    if (opt.value === 'true') {
+                        createOpts.delegate_dataset = true;
+                    } else if (opt.value === 'safe') {
+                        createOpts.delegate_dataset = opt.value;
+                    }
                 }
             }
 
@@ -460,6 +473,22 @@ do_create.options = [
             'cannot be deleted until the protection is disabled. See ' +
             '<https://apidocs.joyent.com/cloudapi/#deletion-protection>'
     },
+    {
+        names: ['delegate-dataset'],
+        type: 'string',
+        help: 'Enable Delegated Datasets on this instance. If TYPE is set to ' +
+              '"true", no additional restrictions are imposed on the way ' +
+              'the instance can use the delegated dataset. If TYPE is set to ' +
+              '"safe", the instance is prevented from receiving ZFS ' +
+              'datasets. Note that Triton CloudAPI instances must have the ' +
+              'SAPI config value ' +
+              '"experimental_cloudapi_delegate_dataset=true" ' +
+              'for create requests using this option to be allowed.',
+        helpArg: 'TYPE',
+        // mark this option hidden for now, see TRITON-853, at least until we
+        // get a full implementation of RFD 044.
+        hidden: true
+    },
     {
         names: ['volume', 'v'],
         type: 'arrayOfString',
